zcash_client_sqlite/src/wallet.rs

```diff
@@ -887,27 +1089,65 @@ fn subtree_scan_progress<P: consensus::Parameters>(
                 )
             })
             .transpose()?
-            // If none of the wallet's accounts have a recover-until height, then we can't
-            // (yet) distinguish general scanning from recovery, so treat the wallet as
-            // fully recovered.
-            .unwrap_or_else(|| Some(Ratio::new(1, 1)));
+            // If none of the wallet's accounts have a recover-until height, then there
+            // is no recovery phase for the wallet, and therefore the denominator in the
+            // resulting ratio (the number of notes in the recovery range) is zero.
+            .unwrap_or_else(|| Some(Ratio::new(0, 0)));
+
         let scan = stmt_scanned_count_from.query_row(
             named_params! {
                 ":start_height": u32::from(
-                    recover_until_height.map(|h| h + 1)
-                        .unwrap_or(birthday_height)
+                    recover_until_height.unwrap_or(birthday_height)
                 ),
             },
             |row| {
                 let scanned = row.get::<_, Option<u64>>(0)?;
                 Ok(scanned.map(|n| Ratio::new(n, n)))
             },
         )?;
-        Ok(Progress { scan, recover })
+
+        Ok(scan.map(|scan| Progress::new(scan, recover)))
     } else {
+        // In case we didn't have information about the tree size at the recover-until
+        // height, get the tree size from a nearby subtree. It's fine for this to be
+        // approximate; it just shifts the boundary between scan and recover progress.
+        let mut get_tree_size_near = |as_of: BlockHeight| {
+            stmt_start_tree_size
+                .query_row(
+                    named_params![":start_height": u32::from(birthday_height)],
+                    |row| row.get::<_, Option<u64>>(0),
+                )
+                .optional()
+                .map(|opt| opt.flatten())
+                .transpose()
+                .or_else(|| {
+                    conn.query_row(
+                        &format!(
+                            "SELECT MIN(shard_index)
+                             FROM {table_prefix}_tree_shards
+                             WHERE subtree_end_height >= :start_height
+                             OR subtree_end_height IS NULL",
+                        ),
+                        named_params! {
+                            ":start_height": u32::from(as_of),
+                        },
+                        |row| {
+                            let min_tree_size = row
+                                .get::<_, Option<u64>>(0)?
+                                .map(|min_idx| min_idx << shard_height);
+                            Ok(min_tree_size)
+                        },
+                    )
+                    .optional()
+                    .map(|opt| opt.flatten())
+                    .transpose()
+                })
+                .transpose()
+        };
```

Comment:
I personally prefer it this way:
```suggestion
        let mut get_tree_size_near = |as_of: BlockHeight| -> Result<_, SqliteClientError> {
            Ok(
                if let Some(tree_size) = stmt_start_tree_size
                    .query_row(
                        named_params![":start_height": u32::from(birthday_height)],
                        |row| row.get::<_, Option<u64>>(0),
                    )
                    .optional()?
                    .flatten()
                {
                    Some(tree_size)
                } else {
                    conn.query_row(
                        &format!(
                            "SELECT MIN(shard_index)
                            FROM {table_prefix}_tree_shards
                            WHERE subtree_end_height >= :start_height
                            OR subtree_end_height IS NULL",
                        ),
                        named_params! {
                            ":start_height": u32::from(as_of),
                        },
                        |row| {
                            let min_tree_size = row
                                .get::<_, Option<u64>>(0)?
                                .map(|min_idx| min_idx << shard_height);
                            Ok(min_tree_size)
                        },
                    )
                    .optional()?
                    .flatten()
                },
            )
        };
```

Reply:
*sigh* it's very frustrating that Rust doesn't have `traverse`, and therefore we have to make do with `.map(...).transpose()`. Personally I find the functional approach nicer, but I'll change it.